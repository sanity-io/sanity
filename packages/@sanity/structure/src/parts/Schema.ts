import {Ordering} from '../Sort'
import {FixMe} from '../types'
import getDefaultModule from './getDefaultModule'

interface Schema {
  name: string
  get(typeName: string): SchemaType | undefined
  getTypeNames(): string[]
}

interface SchemaField {
  name: string
  type: SchemaType
}

interface PreviewFields {
  media?: string
}

interface PreviewPreparer {
  (selection: Record<string, unknown>): PreviewFields
}

export interface SchemaType {
  name: string
  title?: string
  icon?: FixMe
  type?: SchemaType
  to?: SchemaField[]
  fields?: SchemaField[]
  orderings?: Ordering[]
  initialValue?: FixMe | {[key: string]: unknown}
  preview?: {
    select?: PreviewFields
    prepare?: PreviewPreparer
  }
}

// We are lazy-loading the part to work around typescript trying to resolve it
const getDefaultSchema = (): Schema => {
  const schema: Schema = getDefaultModule(require('part:@sanity/base/schema'))
  return schema
}

export {Schema, getDefaultSchema}
