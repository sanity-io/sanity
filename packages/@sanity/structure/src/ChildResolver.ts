import {CollectionBuilder, Collection, SerializePath, SerializeOptions} from './StructureNodes'
import {FixMe} from './types'

export interface ChildResolverOptions {
  index: number
  parent: Collection
  parentPath?: SerializePath
  parameters?: {[key: string]: any}
  serializeOptions?: SerializeOptions
}

export type ItemChild = CollectionBuilder | Collection | FixMe | undefined

interface ChildObservable {
  subscribe: (child: ItemChild | Promise<ItemChild>) => Record<string, unknown>
}

export interface ChildResolver {
  (itemId: string, options: ChildResolverOptions):
    | ItemChild
    | Promise<ItemChild>
    | ChildObservable
    | undefined
}
