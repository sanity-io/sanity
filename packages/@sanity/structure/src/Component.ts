import {camelCase} from 'lodash'
import {SerializeOptions, StructureNode, Serializable, Child} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'
import {validateId} from './util/validateId'
import {FixMe} from './types'

export interface Component extends StructureNode {
  component: FixMe
  child?: Child
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
  options: {[key: string]: unknown}
}

export interface ComponentInput extends StructureNode {
  component: FixMe
  child?: Child
  options?: {[key: string]: unknown}
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export interface BuildableComponent extends Partial<StructureNode> {
  component?: FixMe
  child?: Child
  options?: {[key: string]: unknown}
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export class ComponentBuilder implements Serializable {
  protected spec: BuildableComponent

  constructor(spec?: ComponentInput) {
    this.spec = {options: {}, ...(spec ? spec : {})}
  }

  id(id: string) {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title, id: this.spec.id || camelCase(title)})
  }

  getTitle() {
    return this.spec.title
  }

  child(child: Child) {
    return this.clone({child})
  }

  getChild() {
    return this.spec.child
  }

  component(component: FixMe) {
    return this.clone({component})
  }

  getComponent() {
    return this.spec.component
  }

  options(options: {[key: string]: unknown}) {
    return this.clone({options})
  }

  getOptions() {
    return this.spec.options || {}
  }

  menuItems(menuItems: (MenuItem | MenuItemBuilder)[]) {
    return this.clone({menuItems})
  }

  getMenuItems() {
    return this.spec.menuItems
  }

  menuItemGroups(menuItemGroups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    return this.clone({menuItemGroups})
  }

  getMenuItemGroups() {
    return this.spec.menuItemGroups
  }

  serialize(options: SerializeOptions = {path: []}): Component {
    const {id, title, child, options: componentOptions, component} = this.spec
    if (!id) {
      throw new SerializeError(
        '`id` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!component) {
      throw new SerializeError(
        '`component` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    return {
      id: validateId(id, options.path, options.index),
      title,
      type: 'component',
      child,
      component,
      options: componentOptions || {},
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, options.path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, options.path)
      ),
    }
  }

  clone(withSpec?: BuildableComponent): ComponentBuilder {
    const builder = new ComponentBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
