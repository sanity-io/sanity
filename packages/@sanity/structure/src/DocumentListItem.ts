import {SerializeOptions} from './StructureNodes'
import {Partial} from './Partial'
import {ListItemBuilder, ListItem, UnserializedListItem, ListItemInput} from './ListItem'
import {SchemaType} from './parts/Schema'
import {SerializeError, HELP_URL} from './SerializeError'
import {DocumentBuilder, getDefaultDocumentNode} from './Document'

export interface DocumentListItemInput extends ListItemInput {
  schemaType: SchemaType | string
}

export interface DocumentListItem extends ListItem {
  schemaType: SchemaType
  _id: string
}

type PartialDocumentListItem = Partial<UnserializedListItem>

const getDefaultChildResolver = (spec: PartialDocumentListItem) => (documentId: string) => {
  const schemaType =
    spec.schemaType &&
    (typeof spec.schemaType === 'string' ? spec.schemaType : spec.schemaType.name)

  return schemaType
    ? getDefaultDocumentNode({schemaType, documentId})
    : new DocumentBuilder().id('documentEditor').documentId(documentId)
}

export class DocumentListItemBuilder extends ListItemBuilder {
  protected spec: PartialDocumentListItem

  constructor(spec?: DocumentListItemInput) {
    super(spec)
    this.spec = spec ? spec : {}
  }

  serialize(options: SerializeOptions = {path: []}): DocumentListItem {
    const spec = super.serialize({...options, titleIsOptional: true})

    if (!spec.schemaType) {
      throw new SerializeError(
        '`schemaType` is required for document list items',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.SCHEMA_TYPE_REQUIRED)
    }

    const child = spec.child || getDefaultChildResolver(spec)
    return {...spec, child, schemaType: spec.schemaType, _id: spec.id}
  }

  clone(withSpec?: PartialDocumentListItem) {
    const builder = new DocumentListItemBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
