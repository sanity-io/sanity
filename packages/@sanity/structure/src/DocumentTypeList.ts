import {DocumentListBuilder, DocumentListInput, PartialDocumentList} from './DocumentList'
import {Child} from './StructureNodes'
import {DEFAULT_INTENT_HANDLER} from './Intent'
import {GenericListInput} from './GenericList'
import {SchemaType} from './parts/Schema'

export interface DocumentTypeListInput extends Partial<GenericListInput> {
  schemaType: SchemaType | string
}

export class DocumentTypeListBuilder extends DocumentListBuilder {
  protected spec: PartialDocumentList

  constructor(spec?: DocumentListInput) {
    super()
    this.spec = spec ? spec : {}
  }

  child(child: Child) {
    return this.cloneWithoutDefaultIntentHandler({child})
  }

  clone(withSpec?: PartialDocumentList): DocumentTypeListBuilder {
    const parent = super.clone(withSpec)
    const builder = new DocumentTypeListBuilder()
    builder.spec = {...this.spec, ...parent.getSpec(), ...(withSpec || {})}
    return builder
  }

  cloneWithoutDefaultIntentHandler(withSpec?: PartialDocumentList): DocumentTypeListBuilder {
    const parent = super.clone(withSpec)
    const builder = new DocumentTypeListBuilder()
    const canHandleIntent = this.spec.canHandleIntent
    const shouldOverride = canHandleIntent && canHandleIntent.identity === DEFAULT_INTENT_HANDLER
    const override = shouldOverride ? {canHandleIntent: undefined} : {}
    builder.spec = {
      ...parent.getSpec(),
      ...this.spec,
      ...(withSpec || {}),
      ...override
    }
    return builder
  }
}
