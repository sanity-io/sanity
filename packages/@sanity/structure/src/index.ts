import {uniqueId} from 'lodash'
import {SchemaType} from './parts/Schema'
import {ListBuilder, ListInput} from './List'
import {
  getDocumentTypeListItems,
  getDocumentTypeListItem,
  getDocumentTypeList,
} from './documentTypeListItems'
import {
  MenuItemBuilder,
  MenuItem,
  getOrderingMenuItemsForSchemaType,
  getOrderingMenuItem,
} from './MenuItem'
import {ListItemBuilder, ListItemInput} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {Divider} from './StructureNodes'
import {SerializeError} from './SerializeError'
import {
  DocumentBuilder,
  PartialDocumentNode,
  documentFromEditor,
  documentFromEditorWithInitialValue,
  getDefaultDocumentNode,
} from './Document'
import {ComponentInput, ComponentBuilder} from './Component'
import {DocumentListItemBuilder, DocumentListItemInput} from './DocumentListItem'
import {Ordering} from './Sort'
import * as views from './views'
import {
  InitialValueTemplateItemBuilder,
  defaultInitialValueTemplateItems,
  menuItemsFromInitialValueTemplateItems,
} from './InitialValueTemplateItem'

const StructureBuilder = {
  defaults: getDefaultStructure,
  documentTypeList: getDocumentTypeList,
  documentTypeListItem: getDocumentTypeListItem,
  documentTypeListItems: getDocumentTypeListItems,
  document: (spec?: PartialDocumentNode) => new DocumentBuilder(spec),
  documentWithInitialValueTemplate: documentFromEditorWithInitialValue,
  defaultDocument: getDefaultDocumentNode,

  list: (spec?: ListInput) => new ListBuilder(spec),
  listItem: (spec?: ListItemInput) => new ListItemBuilder(spec),

  menuItem: (spec?: MenuItem) => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup) => new MenuItemGroupBuilder(spec),
  menuItemsFromInitialValueTemplateItems,

  documentList: (spec?: DocumentListInput) => new DocumentListBuilder(spec),
  documentListItem: (spec?: DocumentListItemInput) => new DocumentListItemBuilder(spec),

  orderingMenuItem: (ordering: Ordering) => getOrderingMenuItem(ordering),
  orderingMenuItemsForType: (type: string) => getOrderingMenuItemsForSchemaType(type),

  editor: documentFromEditor,
  editorWithInitialValueTemplate: documentFromEditorWithInitialValue,

  defaultInitialValueTemplateItems,
  initialValueTemplateItem: (
    templateId: string,
    parameters?: {[key: string]: any}
  ): InitialValueTemplateItemBuilder =>
    new InitialValueTemplateItemBuilder({
      id: templateId,
      parameters,
      templateId,
    }),

  component: (spec?: ComponentInput | Function) => {
    return typeof spec === 'function'
      ? new ComponentBuilder().component(spec)
      : new ComponentBuilder(spec)
  },

  divider: (): Divider => ({id: uniqueId('__divider__'), type: 'divider'}),

  view: views,
}

function hasIcon(schemaType?: SchemaType | string): boolean {
  if (!schemaType || typeof schemaType === 'string') {
    return false
  }

  return Boolean(schemaType.icon)
}

function getDefaultStructure(): ListBuilder {
  const items = getDocumentTypeListItems()
  return new ListBuilder()
    .id('__root__')
    .title('Content')
    .items(items)
    .showIcons(items.some((item) => hasIcon(item.getSchemaType())))
}

export type {InitialValueTemplateItem} from './InitialValueTemplateItem'

export {StructureBuilder, SerializeError}
