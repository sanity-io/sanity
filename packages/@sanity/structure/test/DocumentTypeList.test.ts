import {StructureBuilder as S} from '../src'
import serializeStructure from './util/serializeStructure'
import {getDefaultSchema, SchemaType} from '../src/parts/Schema'

test('builds document type lists with only required properties', () => {
  expect(
    S.documentTypeList({
      id: 'custom-id',
      title: 'Custom author title',
      schemaType: 'author',
    }).serialize({
      path: [],
    })
  ).toMatchSnapshot()
})

test('builds document type lists with schema type as string', () => {
  expect(
    S.documentTypeList('author').serialize({
      path: [],
    })
  ).toMatchSnapshot()
})

test('builds document type lists with schema type name + schema', () => {
  expect(
    S.documentTypeList('author', getDefaultSchema()).serialize({
      path: [],
    })
  ).toMatchSnapshot()
})

test('builds document type lists with schema type instance', () => {
  expect(
    S.documentTypeList({
      schemaType: getDefaultSchema().get('author') as SchemaType,
    }).serialize({
      path: [],
    })
  ).toMatchSnapshot()
})

test('throws if no filter is set', () => {
  expect(() =>
    S.documentTypeList('author').id('foo').filter('').serialize()
  ).toThrowErrorMatchingSnapshot()
})

test('builds document type lists through setters', () => {
  expect(
    S.documentTypeList('author')
      .id('authors')
      .title('authors')
      .filter('_type == $type')
      .params({type: 'author'})
      .defaultLayout('card')
      .defaultOrdering([{field: 'title', direction: 'asc'}])
      .serialize()
  ).toMatchSnapshot()
})

test('builds document type lists through setters (alt order)', () => {
  expect(
    S.documentTypeList('author')
      .defaultOrdering([{field: 'title', direction: 'desc'}])
      .id('authors')
      .title('authors')
      .filter('_type == $type')
      .params({type: 'author'})
      .serialize()
  ).toMatchSnapshot()
})

test('builds document type lists through setters (alt order #2)', () => {
  expect(
    S.documentTypeList('author')
      .params({type: 'author'})
      .defaultOrdering([{field: 'title', direction: 'desc'}])
      .id('authors')
      .title('authors')
      .filter('_type == $type')
      .serialize()
  ).toMatchSnapshot()
})

test('default child resolver resolves to editor', (done) => {
  const list = S.documentTypeList('author').serialize()

  const context = {parent: list, index: 1}
  serializeStructure(list.child, context, ['asoiaf-wow', context]).subscribe((child) => {
    expect(child).toEqual({
      child: undefined,
      id: 'documentEditor',
      type: 'document',
      options: {
        id: 'asoiaf-wow',
        type: 'author',
      },
      views: [
        {
          id: 'editor',
          title: 'Editor',
          type: 'form',
          icon: undefined,
        },
      ],
    })
    done()
  })
})

test('builder is immutable', () => {
  const original = S.documentTypeList('author')
  expect(original.id('foo')).not.toBe(original)
  expect(original.title('foo')).not.toBe(original)
  expect(original.filter('foo == "bar"')).not.toBe(original)
  expect(original.params({foo: 'bar'})).not.toBe(original)
  expect(original.menuItems([])).not.toBe(original)
  expect(original.showIcons(false)).not.toBe(original)
  expect(original.menuItemGroups([])).not.toBe(original)
  expect(original.defaultLayout('card')).not.toBe(original)
  expect(original.child(() => undefined)).not.toBe(original)
  expect(original.canHandleIntent(() => false)).not.toBe(original)
  expect(original.defaultOrdering([{field: 'title', direction: 'asc'}])).not.toBe(original)
})

test('getters work', () => {
  const original = S.documentTypeList('author')
  const child = () => undefined
  const canHandleIntent = () => false
  const field = 'title'
  const direction = 'asc'
  expect(original.id('foo').getId()).toEqual('foo')
  expect(original.title('bar').getTitle()).toEqual('bar')
  expect(original.filter('foo == "bar"').getFilter()).toEqual('foo == "bar"')
  expect(original.params({foo: 'bar'}).getParams()).toEqual({foo: 'bar'})
  expect(original.menuItems([]).getMenuItems()).toEqual([])
  expect(original.menuItemGroups([]).getMenuItemGroups()).toEqual([])
  expect(original.defaultLayout('card').getDefaultLayout()).toEqual('card')
  expect(original.child(child).getChild()).toEqual(child)
  expect(original.showIcons(false).getShowIcons()).toEqual(false)
  expect(original.canHandleIntent(canHandleIntent).getCanHandleIntent()).toEqual(canHandleIntent)
  expect(original.defaultOrdering([{field, direction}]).getDefaultOrdering()).toEqual([
    {field, direction},
  ])
})

test('can disable icons from being displayed', () => {
  const list = S.documentTypeList('author')
    .id('blamuggost')
    .title('Bl√•muggost')
    .filter('_type == "bluecheese"')
    .showIcons(false)

  expect(list.serialize()).toMatchObject({
    id: 'blamuggost',
    displayOptions: {showIcons: false},
  })

  expect(list.getShowIcons()).toBe(false)
})
