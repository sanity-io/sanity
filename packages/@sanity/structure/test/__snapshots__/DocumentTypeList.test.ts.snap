// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds document type lists through setters (alt order #2) 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "authors",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "title",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "authors",
  "type": "documentList",
}
`;

exports[`builds document type lists through setters (alt order) 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "authors",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "title",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "authors",
  "type": "documentList",
}
`;

exports[`builds document type lists through setters 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": "card",
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "authors",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "asc",
        "field": "title",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "authors",
  "type": "documentList",
}
`;

exports[`builds document type lists with only required properties 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "custom-id",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "_updatedAt",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "Custom author title",
  "type": "documentList",
}
`;

exports[`builds document type lists with schema type as string 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "author",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "_updatedAt",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "Author",
  "type": "documentList",
}
`;

exports[`builds document type lists with schema type instance 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "author",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "_updatedAt",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "Author",
  "type": "documentList",
}
`;

exports[`builds document type lists with schema type name + schema 1`] = `
Object {
  "canHandleIntent": [Function],
  "child": [Function],
  "defaultLayout": undefined,
  "displayOptions": Object {
    "showIcons": true,
  },
  "id": "author",
  "initialValueTemplates": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "sorting",
      "title": "Sort",
    },
    Object {
      "id": "layout",
      "title": "Layout",
    },
    Object {
      "id": "actions",
      "title": "Actions",
    },
  ],
  "menuItems": Array [
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_updatedAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Last edited",
    },
    Object {
      "action": "setSortOrder",
      "group": "sorting",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "by": Array [
          Object {
            "direction": "desc",
            "field": "_createdAt",
          },
        ],
        "extendedProjection": "",
      },
      "title": "Sort by Created",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "default",
      },
      "title": "Compact view",
    },
    Object {
      "action": "setLayout",
      "group": "layout",
      "icon": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "params": Object {
        "layout": "detail",
      },
      "title": "Detailed view",
    },
  ],
  "options": Object {
    "apiVersion": "2021-06-07",
    "defaultOrdering": Array [
      Object {
        "direction": "desc",
        "field": "_updatedAt",
      },
    ],
    "filter": "_type == $type",
    "params": Object {
      "type": "author",
    },
  },
  "schemaTypeName": "author",
  "title": "Author",
  "type": "documentList",
}
`;

exports[`throws if no filter is set 1`] = `"\`filter\` is required for document lists"`;
